server/src/remote.rs

```diff
@@ -296,19 +304,236 @@ impl FlightService for FlightServiceHandler {
         Err(Status::unimplemented("Not implemented."))
     }
 
-    /// Not implemented.
+    /// Perform a specific action based on the type of the action in `request`. Currently supports
+    /// two actions: `CreateTable` and `CreateModelTable`. `CreateTable` creates a normal table
+    /// in the catalog when given a table name and schema. `CreateModelTable` creates a table
+    /// that can be used for ingestion when given a table name, a schema, a list of indices
+    /// specifying which columns are metadata tag columns, and an index specifying which column
+    /// is the timestamp column.
+    ///
+    /// The data is given in the action body and must have the following format:
+    /// The first two bytes are the length x of the first argument. The next x bytes are the first
+    /// argument. This pattern repeats until all arguments are consumed.
     async fn do_action(
         &self,
-        _request: Request<Action>,
+        request: Request<Action>,
     ) -> Result<Response<Self::DoActionStream>, Status> {
-        Err(Status::unimplemented("Not implemented."))
+        let action = request.into_inner();
+        info!("Received request to perform action '{}'.", action.r#type);
+
+        if action.r#type == "CreateTable" || action.r#type == "CreateModelTable" {
+            // Extract the table name from the action body.
+            let (table_name_bytes, table_name_offset) = extract_argument_bytes(&action.body);
+            let table_name = str::from_utf8(table_name_bytes)
+                .map_err(|error| Status::invalid_argument(error.to_string()))?;
+
+            // Check if the table name is a valid object_store path and database table name.
+            object_store::path::Path::parse(table_name)
+                .map_err(|error| Status::invalid_argument(error.to_string()))?;
+
+            if table_name.contains(char::is_whitespace) {
+                return Err(Status::invalid_argument("Table name cannot contain whitespace.".to_owned()));
+            }
+
+            // Extract the schema from the action body.
+            let offset_data = &action.body[table_name_offset..];
+            let (schema_bytes, schema_offset) = extract_argument_bytes(offset_data);
+            let ipc_message = IpcMessage(Vec::from(schema_bytes));
+            let schema = Schema::try_from(ipc_message)
+                .map_err(|error| Status::invalid_argument(error.to_string()))?;
+
+            // Get the write lock for the catalog to ensure that multiple writes cannot happen at the same time.
+            let catalog = self.context.catalog.write().unwrap();
```

Comment:
I have added the text to explain why `unwrap()` is safe. I have decided not to move the `write()` call into the two `if` branches since the create table actions should be called rarely and to avoid duplicated code.